### PR TITLE
[DomCrawler] Encode html entities only if nessecary

### LINKED ISSUE #54437: DomCrawler encodes html entities inside script tags
### Symfony version(s) affected

5.4.37

### Description

The `Crawler` encodes html entities before parsing it in https://github.com/symfony/symfony/blob/f28e58eef622a4524b8ada8ddf252130a39094fa/src/Symfony/Component/DomCrawler/Crawler.php#L1184 which causes the contents of script tags to be corrupt (because HTML entities have no meaning between `<script>` and `</script>`).

### How to reproduce

```php
$crawler = new Crawler();
$crawler->addContent('<!doctype html><html><script>var foo = "bär";</script></html>', 'text/html; charset=UTF-8');
echo $crawler->filterXPath('//script')->text();
// output: var foo = "b&#228;r";
// expected: var foo = "bär";
```

### Possible Solution

I’m not sure what’s the best way to fix it as `convertToHtmlEntities()` cannot distinguish between outside and inside `<script>`. But as it seems that `convertToHtmlEntities()` is only there to fix issues with the libxml parser, maybe it can be skipped if the `Masterminds\HTML5` is used?

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by ausi**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #54437
| License       | MIT

As discussed in https://github.com/symfony/symfony/issues/54437#issuecomment-2027083859

### CODE REVIEW

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 1]
```diff
@@ -1151,12 +1151,30 @@ protected function sibling(\DOMNode $node, string $siblingDir = 'nextSibling')
...
+            return '' === @mb_convert_encoding('', $encoding, 'UTF-8');
+        } catch (\Throwable $e) {
+            return false;
+        }
     }
 
     private function parseXhtml(string $htmlContent, string $charset = 'UTF-8'): \DOMDocument
     {
-        $htmlContent = $this->convertToHtmlEntities($htmlContent, $charset);
+        if ('UTF-8' === $charset && 1 === preg_match('//u', $htmlContent)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DomCrawler/Crawler.php]:
```suggestion
        if ('UTF-8' === $charset && preg_match('//u', $htmlContent)) {
```

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 2]
```diff
@@ -1151,12 +1151,30 @@ protected function sibling(\DOMNode $node, string $siblingDir = 'nextSibling')
...
+
+        return $this->html5Parser->parse($htmlContent, ['encoding' => $charset]);
+    }
+
+    private function supportsEncoding(string $encoding): bool
+    {
+        try {
+            return '' === @mb_convert_encoding('', $encoding, 'UTF-8');
+        } catch (\Throwable $e) {
+            return false;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DomCrawler/Crawler.php:1167]: can this code path be tested? is it already?

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 3]
```diff
@@ -1151,12 +1151,30 @@ protected function sibling(\DOMNode $node, string $siblingDir = 'nextSibling')
...
+        }
     }
 
     private function parseXhtml(string $htmlContent, string $charset = 'UTF-8'): \DOMDocument
     {
-        $htmlContent = $this->convertToHtmlEntities($htmlContent, $charset);
+        if ('UTF-8' === $charset && 1 === preg_match('//u', $htmlContent)) {
+            $htmlContent = '<?xml encoding="UTF-8">'.$htmlContent;
+        } else {
+            $htmlContent = $this->convertToHtmlEntities($htmlContent, $charset);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DomCrawler/Crawler.php:1176]: can this code path be tested? is it already?

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 2 above)
Comment (ausi) [reply src/Symfony/Component/DomCrawler/Crawler.php:1167]: Yes, the catch is tested (PHP 8) and the `'' === ` part is tested on PHP 7. Both via the `Html5ParserCrawlerTest::testAddHtmlContentUnsupportedCharset()`

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 3 above)
Comment (ausi) [reply src/Symfony/Component/DomCrawler/Crawler.php:1176]: Yes, this is tested by `NativeParserCrawlerTest::testAddHtmlContentCharsetGbk()`.

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ausi.

**Pull request merged**